### ISSUE TITLE
Identifying VPN Servers through Graph-Represented Behaviors (ACM WWW 2024)

### ISSUE THREAD

**Issue by wkrp** [net4people/bbs#455]
Identifying VPN Servers through Graph-Represented Behaviors
王晨旭 (Wang Chenxu), 尹姜谊 (Yin Jiangyi), 李钊 (Li Zhao), 徐红博 (Xu Hongbo), 张中一 (Zhang Zhongyi), 刘庆云 (Liu Qingyun)
https://dl.acm.org/doi/10.1145/3589334.3645552
[PDF](https://github.com/user-attachments/files/18975619/3589334.3645552.pdf)
https://github.com/chenxuStep/VPNChecker Partial source code and data

The peer reviews for this paper are public and you can read them at https://openreview.net/forum?id=7024czziih.

The paper presents a machine learning classifier to classify network servers as VPN or non-VPN, using a combination of active probing and observed communication between endpoints. There's a preliminary assumption that you start with a log of client–server access records (which in this case are provided by an ISP). The technique is based around building two graphs. In the *probing graph*, vertices represent servers, and edges connect pairs of servers that have similar patterns of open TCP or UDP ports. The *communication graph* straightforwardly represents observed access patterns: vertices are clients and servers, and an edge between two hosts means they communicated at some point. In both graphs, vertices have an associated feature vector. To classify a server, you use [GraphSAGE](https://snap.stanford.edu/graphsage/) vertex embedding, which produces an aggregated feature vector from a vertex's own features and the features of its local neighborhood. You concatenate the aggregated feature vectors of the server of interest from the probing and communication graphs, then pass the long feature vector into an ordinary neural network classifier.

The final classification step is ordinary machine learning classification over a linear feature vector. The "graph-represented behaviors" of the title refer not only to client–server access patterns (which have an obvious graph representation), but also server features derived from active probing (which get represented as a more abstract graph). The intuition behind the communication graph is that it can help with classifying [probe-resistant servers](https://censorbib.nymity.ch/#Frolov2020a). The paper says that VPN servers tend to be accessed by fewer distinct clients, and have overall less "sophisticated" connection relations than non-VPN servers (but does not really analyze this claim).

The first graph, the probing graph, mostly reuses past work on probing-based VPN identification, such as ["Detecting Probe-resistant Proxies"](https://censorbib.nymity.ch/#Frolov2020a), ["ACER"](https://link.springer.com/article/10.1007/s11416-020-00353-z), and ["OpenVPN is Open to VPN Fingerprinting"](https://censorbib.nymity.ch/#Xue2022a). But there are some new ideas, such as linking servers that have similar patterns of open ports, and finding "stealth ports". You start by doing TCP and UDP port scans of all known servers to find open ports. Then, you join two server vertices by an edge whenever the [Jaccard similarity](https://en.wikipedia.org/wiki/Jaccard_similarity) of their respective sets of open ports exceeds a threshold <var>ζ</var>. (In the evaluation, they set <var>ζ</var> = 0.5.) The authors say that certain patterns of open ports are characteristic of certain VPNs; see [Table 6](https://github.com/user-attachments/files/18975619/3589334.3645552.pdf#page=10). Besides the edges between vertices, each vertex has an associated feature vector, which includes such features as:

- Codes indicating responses to [application-specific probes](https://github.com/chenxuStep/VPNChecker/blob/309101af6fde2cec763df66fe53686407b481dee/code/probes.py) (SSTP, IPsec, OpenVPN, PPTP, FTP, SSH, DNS, HTTP, TLS) sent to standard ports. [Response type codes](https://github.com/chenxuStep/VPNChecker/blob/309101af6fde2cec763df66fe53686407b481dee/code/responseType.py) indicate conditions such as the payload matching a pattern, RST, timeout, and others.
- Packet timing, length, and count features from responses to MTU-sized TCP and UDP probes. The authors say that VPN servers tend to reply with shorter packets, and longer delays, than non-VPN servers.
- TCP RST/FIN response threshold, found through binary search.
- Counts of passively observed ports, actively probed open ports, and "stealth ports". Stealth ports are server ports that were observed communicating in passive access logs, but which did not respond when actively probed. Stealth ports are consistent with authenticated, probe-resistant protocols like Shadowsocks or obfs4.
- A compressed representation of the server's set of open ports.

The second graph, the communication graph, represents communication between hosts in the access log. Each vertex in the communication graph is a client IP address, server IP address, or domain name. (Or so I believe—this part of the paper is ambiguous. They apparently augment the IP addresses in the access log with domain names from DNS PTR records and TLS certificates.) If two hosts ever communicated in the access log, their corresponding vertices are connected by an edge in the communication graph. As in the probing graph, each vertex in the communication graph has an associated feature vector. The features here have to do with graph topology: degree, eigenvector centrality, PageRank, closeness centrality, local clustering coefficient, and <var>k</var>-core.

The evaluation is problematic, both because it's not clear that it's representative, and because the reported performance is not very good. The evaluation is divided into an "offline" training using a recorded access log from an ISP, and an "online" test of live deployment on the same ISP. The central problem is the lack of ground-truth classification labels for VPN and non-VPN servers. For the offline part, the authors got an access log containing 46,481 servers. The labeling heuristic they apply is pretty crude: any IP address in [AS 9009](https://stat.ripe.net/ui2013/widget/whois#w.resource=AS9009) (M247) or [AS 60068](https://stat.ripe.net/ui2013/widget/whois#w.resource=AS60068) (CDN77) is considered a VPN server; all others are considered non-VPN. They additionally get labels for specific VPNs (Psiphon3, NordVPN, Surfshark, etc.) from threat intelligence provided by an unnamed industry partner. (This part of the paper is quite unclear. They call the heuristically labeled VPN servers "seed IPs", but it does not appear they use them to expand the scope of known VPN servers, as you would expect from the name. The seed IPs instead go into something called "server sequence access length <var>ρ</var>", which is not adequately explained.) The labeling heuristic for the online part of the evaluation is different for some reason, and similarly questionable. Here they have 512,710 server IP addresses to deal with. They do things like grep hostnames for strings such as "vpn" and "surfshark", send additional active probes, and consult third-party VPN classification services. In essence, the evaluation presents two classifiers of unknown quality (namely, VPNChecker and various heuristics) and tests how well they agree with each other, which is not very meaningful. Reported accuracy, precision, recall, and F<sub>1</sub> scores are in the 85%–90% range, and not always better than prior work.

The authors unfortunately do not spend much time analyzing what it is about certain features that make them distinguish VPNs. (Beyond what is already analyzed in the prior work that they reuse.) The insights offered are:

- VPN servers tend to have fewer distinct users and simpler access patterns than non-VPN servers.
- VPN servers have characteristic combinations of open ports, such as {22, 53, 443}, {22, 443, 554}, and {22, 53, 443, 554} for Psiphon ([Table 6](https://github.com/user-attachments/files/18975619/3589334.3645552.pdf#page=10)).
- VPN servers are more likely to have "stealth ports".

**Comment by Hawzen** [comment]
> The evaluation is problematic, both because it's not clear that it's representative, and because the reported performance is not very good. The evaluation is divided into an "offline" training using a recorded access log from an ISP, and an "online" test of live deployment on the same ISP. The central problem is the lack of ground-truth classification labels for VPN and non-VPN servers. For the offline part, the authors got an access log containing 46,481 servers. The labeling heuristic they apply is pretty crude: any IP address in [AS 9009](https://stat.ripe.net/ui2013/widget/whois#w.resource=AS9009) (M247) or [AS 60068](https://stat.ripe.net/ui2013/widget/whois#w.resource=AS60068) (CDN77) is considered a VPN server; all others are considered non-VPN. They additionally get labels for specific VPNs (Psiphon3, NordVPN, Surfshark, etc.) from threat intelligence provided by an unnamed industry partner. (This part of the paper is quite unclear. They call the heuristically labeled VPN servers "seed IPs", but it does not appear they use them to expand the scope of known VPN servers, as you would expect from the name. The seed IPs instead go into something called "server sequence access length ρ", which is not adequately explained.) The labeling heuristic for the online part of the evaluation is different for some reason, and similarly questionable. Here they have 512,710 server IP addresses to deal with. They do things like grep hostnames for strings such as "vpn" and "surfshark", send additional active probes, and consult third-party VPN classification services. In essence, the evaluation presents two classifiers of unknown quality (namely, VPNChecker and various heuristics) and tests how well they agree with each other, which is not very meaningful. Reported accuracy, precision, recall, and F1 scores are in the 85%–90% range, and not always better than prior work.

If we were to repeat their experiment, what's the optimal way of obtaining ground truth? 

This is a very difficult problem, so I'm curious what do you guys think

**Comment by wkrp** [comment]
> If we were to repeat their experiment, what's the optimal way of obtaining ground truth?

If you're starting from a list of IP addresses that are just given to you, as in this paper, I'm not sure there is a good way. That's because it would be begging the question. Suppose you had a high-confidence way to distinguish VPN servers, reliable enough to establish ground truth for the evaluation of your new machine learning classifier. Well, then, there's no need for your new machine learning classifier, because you already have something better.

What would be somewhat more principled is to start with a set of *known* VPN/non-VPN servers, train a classifier on it, and then hope the classifier generalizes to *other* servers that were not in the training set. The risk there, of course, is that you don't know what you don't know: if there are significant classes of VPN or non-VPN servers that you didn't think to include in the training set, then the classifier is likely to misclassify them. But really, the same risk exists with the approach taken in this paper. Anything that their labeling heuristics mislabel is likely to be misclassified by the actual classifier. The approach taken in this paper leads to a classifier that predicts the output of the labeling heuristics, which is not necessarily the same as whether the server is a VPN server or not.

I'm not au fait with this line of research, but the references in this paper (particularly in Section 4.3) show some of the ways that VPN inference has been done in research.

* ["VPNalyzer: Systematic Investigation of the VPN Ecosystem"](https://www.ndss-symposium.org/ndss-paper/auto-draft-244/), Table IV, is the source of the AS 9009/60068 observation in Section 4.1. The way it's used in this paper is appropriate, as the VPNalyzer paper makes neither the claim that all VPNs are in those two ASes, nor that all non-VPNs are outside those two ASes.
* ["Characterizing the VPN Ecosystem in the Wild"](https://vpnecosystem.github.io/)
* ["The Lockdown Effect: Implications of the COVID-19 Pandemic on Internet Traffic"](https://dl.acm.org/doi/10.1145/3419394.3423658) Section 6

**Comment by seeker252** [comment]
Hello.
Today morning I thought, that I found some serious Reality volnurability. I hope to listen from you that I mistaken.
Look. Reality works as TLS 1.3 server and use SessionID for authenticate valid client connections. All requests from other clients transparently redirected to remote server by TCP. In TLS 1.3 SessionID must be random ir zero. Any TLS 1.3 server don't validate it. Ok. If third party (censor) change Reality client SessionID (with encrypted authentication data) in ClientHello with any 32 bit data for all TLS 1.3 Hellos to foreign IPs or IPs from grey lists, I think all legal (non-Reality) TLS 1.3 connections not be violated. We even can connect with our Reality VPS, but Reality server due to Session ID changed by third party will think about us as not valid Reality client and send our connection to remote server.
So censor no need to find Reality servers specially. He may mechanically change Session ID of TLS 1.3 ClientHello to any randon 32-bit value for some connections to some IPs and not violate any TLS 1.3 non-Reality connections. So HTTPs will work fine, but Reality don't.
Or I mistaken?

**Comment by wkrp** [comment]
The [Geedge/MESA leak](https://github.com/net4people/bbs/issues/519) contains a presentation that talks about the peer review process of this paper, among other things.

[mesalab_docs/kuailekeyan/组内写作经验&参会分享.html](https://geedge-docs.haruue.com/mesalab_docs/kuailekeyan/%E7%BB%84%E5%86%85%E5%86%99%E4%BD%9C%E7%BB%8F%E9%AA%8C&%E5%8F%82%E4%BC%9A%E5%88%86%E4%BA%AB.html)

> 组内写作经验&参会分享

> *Writing experience within the group & sharing at the meeting*

[mesalab_docs/kuailekeyan/attachments/106110515_attachments_王晨旭_论文撰写及参会交流分享.pdf](https://geedge-docs.haruue.com/mesalab_docs/kuailekeyan/attachments/106110515_attachments_%E7%8E%8B%E6%99%A8%E6%97%AD_%E8%AE%BA%E6%96%87%E6%92%B0%E5%86%99%E5%8F%8A%E5%8F%82%E4%BC%9A%E4%BA%A4%E6%B5%81%E5%88%86%E4%BA%AB.pdf)

[Page 9](https://geedge-docs.haruue.com/mesalab_docs/kuailekeyan/attachments/106110515_attachments_%E7%8E%8B%E6%99%A8%E6%97%AD_%E8%AE%BA%E6%96%87%E6%92%B0%E5%86%99%E5%8F%8A%E5%8F%82%E4%BC%9A%E4%BA%A4%E6%B5%81%E5%88%86%E4%BA%AB.pdf#page=9):

> ### WWW2024-VPNTracker论文写作解构
>
> * 背景：VPN Server检测
> * Idea来源：对线上数据的观察。
>   * 发现VPN厂商共享端口组合 (线上用于确定候选集)
>   * 当观测到客户端访问VPN IP时，在时间窗口T内，大概率还会发现其他VPN IP（额外的实验）
> * 论文撰写阶段一 场景构建：
>   * 基于Client IP – Server IP – Port 之间的关系，利用图神经网络进行识别
>   * 特征：线上日志数据，流持续时间、Server IP访问数量…
>   * 结果：有效特征比较少，分类效果非常差。
> * 论文撰写阶段二 特征处理：
> * 复现 USENIX2022、NDSS2020…相关论文的工作

> ### *WWW2024-VPNTracker Paper Writing Deconstruction*
>
> * *Background: VPN Server detection*
> * *Source of Idea: Observation of online data.*
>   * *Discover VPN vendors' shared port combinations (used online to determine candidate sets)*
>   * *When it is observed that the client accesses the VPN IP, there is a high probability that other VPN IPs will be found within the time window T (additional experiments)*
> * *Paper writing stage 1 Scenario construction:*
>   * *Based on the relationship between Client IP – Server IP – Port, use graph neural network for identification*
>   * *Features: Online log data, flow duration, number of Server IP visits…*
>   * *Result: There are relatively few effective features and the classification effect is very poor*
> * *Paper writing stage 2 feature processing:*
> * *Reproduce the work of USENIX2022, NDSS2020, … related papers*
